samples/matter/bridge/src/bridge_manager.cpp

```diff
@@ -0,0 +1,392 @@
+
+CHIP_ERROR BridgeManager::AddBridgedDevice(BridgedDevice::DeviceType bridgedDeviceType, char *nodeLabel)
+{
+	BridgedDevice *device;
+
+	char *label = nullptr;
+	if (nodeLabel && (strlen(nodeLabel) >= BridgedDevice::kNodeLabelSize)) {
+		return CHIP_ERROR_INVALID_ARGUMENT;
+	}
+
+	if (nodeLabel) {
+		label = nodeLabel;
+	}
+
+	switch (bridgedDeviceType) {
+#if CONFIG_BRIDGE_ONOFF_LIGHT_BRIDGED_DEVICE
+	case BridgedDevice::DeviceType::OnOffLight: {
+		LOG_INF("Adding OnOff Light bridged device");
+		device = reinterpret_cast<BridgedDevice *>(Platform::New<OnOffLightDevice>(label));
+		break;
+	}
+#endif /* CONFIG_BRIDGE_ONOFF_LIGHT_BRIDGED_DEVICE */
+#if CONFIG_BRIDGE_TEMPERATURE_SENSOR_BRIDGED_DEVICE
+	case BridgedDevice::DeviceType::TemperatureSensor: {
+		LOG_INF("Adding TemperatureSensor bridged device");
+		device = reinterpret_cast<BridgedDevice *>(Platform::New<TemperatureSensorDevice>(label));
+		break;
+	}
+#endif /* CONFIG_BRIDGE_TEMPERATURE_SENSOR_BRIDGED_DEVICE */
+#if CONFIG_BRIDGE_HUMIDITY_SENSOR_BRIDGED_DEVICE
+	case BridgedDevice::DeviceType::HumiditySensor: {
+		LOG_INF("Adding HumiditySensor bridged device");
+		device = reinterpret_cast<BridgedDevice *>(Platform::New<HumiditySensorDevice>(label));
+		break;
+	}
+#endif /* CONFIG_BRIDGE_HUMIDITY_SENSOR_BRIDGED_DEVICE */
+	default:
+		return CHIP_ERROR_INVALID_ARGUMENT;
+	}
+
+	if (!device) {
+		return CHIP_ERROR_NO_MEMORY;
+	}
+
+	return AddDevice(device);
+}
+
+CHIP_ERROR BridgeManager::RemoveBridgedDevice(uint16_t endpoint)
+{
+	uint8_t index = 0;
+
+	while (index < kMaxBridgedDevices) {
+		if (nullptr != mBridgedDevices[index]) {
+			if (mBridgedDevices[index]->GetEndpointId() == endpoint) {
+				LOG_INF("Removed dynamic endpoint %d (index=%d)", endpoint, index);
+				/* Free dynamically allocated memory */
+				emberAfClearDynamicEndpoint(index);
+				Platform::Delete(mBridgedDevices[index]);
+				mBridgedDevices[index] = nullptr;
+				return CHIP_NO_ERROR;
+			}
+		}
+		index++;
+	}
+
+	return CHIP_ERROR_NOT_FOUND;
+}
+
+CHIP_ERROR BridgeManager::AddDevice(BridgedDevice *device)
+{
+	uint8_t index = 0;
+
+	while (index < kMaxBridgedDevices) {
+		/* Find the first empty index in the brdiged devices list */
+		if (nullptr == mBridgedDevices[index]) {
+			mBridgedDevices[index] = device;
+			EmberAfStatus ret;
+			while (true) {
+				ret = emberAfSetDynamicEndpoint(
+					index, mCurrentDynamicEndpointId, device->mEp,
+					Span<DataVersion>(device->mDataVersion, device->mDataVersionSize),
+					Span<const EmberAfDeviceType>(device->mDeviceTypeList,
+								      device->mDeviceTypeListSize));
+
+				if (ret == EMBER_ZCL_STATUS_SUCCESS) {
+					LOG_INF("Added device to dynamic endpoint %d (index=%d)",
+						mCurrentDynamicEndpointId, index);
+					device->Init(mCurrentDynamicEndpointId);
+					return CHIP_NO_ERROR;
+				} else if (ret != EMBER_ZCL_STATUS_DUPLICATE_EXISTS) {
+					LOG_ERR("Failed to add dynamic endpoint: Internal error!");
+					return CHIP_ERROR_INTERNAL;
```

Comment:
Then we keep `mBridgedDevices[index]` allocated?

Reply:
I think that we could move `mBridgedDevices[index] = device;` to the success case. Moreover I think that the output from `AddDevice` should be verified and in case of failure we should `Platform::Delete` the allocated memory for `device`.

Reply:
Right, that makes sense.